### PR TITLE
Keep the existing default catalog for `SessionStateBuilder::new_from_existing`

### LINKED ISSUE #11988: Disable `create_default_catalog` when the exist session state has default catalog for `SessionStateBuilder`
### Is your feature request related to a problem or challenge?

I found that `SessionStateBuilder::new_from_existing` will unset the default catalog of the existing state. Consider the following case:
```rust
    fn employee_batch() -> RecordBatch {
        let name: ArrayRef = Arc::new(StringArray::from_iter_values([
            "Andy",
            "Andrew",
            "Oleks",
            "Chunchun",
            "Xiangpeng",
        ]));
        let age: ArrayRef = Arc::new(Int32Array::from(vec![11, 22, 33, 44, 55]));
        let position = Arc::new(StringArray::from_iter_values([
            "Engineer", "Manager", "Engineer", "Manager", "Engineer",
        ]));
        RecordBatch::try_from_iter(vec![("name", name), ("age", age), ("position", position)])
            .unwrap()
    }

    let ctx = SessionContext::new();
    ctx.register_batch("employee", employee_batch())?;
    let table = ctx.catalog("datafusion").unwrap().schema("public").unwrap().table("employee").await?;
    println!("{}", table.is_some());
    let new_state = SessionStateBuilder::new_from_existing(ctx.state()).build();
    let table = new_state.catalog_list().catalog("datafusion").unwrap().schema("public").unwrap().table("employee").await?;
    println!("{}", table.is_some());
```
The output result is
```
true
false
```

This behavior confuses me. After some research, I found that we have a configuration option, `create_default_catalog_and_schema`, which is true by default. However, I think the user might expect the new one to be exactly the same as the existing one, except for the session ID.

### Describe the solution you'd like

I plan to add a check for `SessionStateBuilder::new_from_existing`. If the default catalog exists in the existing state, we can disable `create_default_catalog_and_schema` by default. 

### Describe alternatives you've considered

If we don't make this change, I think we should enhance the documentation for `SessionStateBuilder::new_from_existing`. Currently, the documentation only mentions that the session ID will be unset, while other settings remain the same.

https://github.com/apache/datafusion/blob/e4be013064943786c9915bbc79c18ee82106340a/datafusion/core/src/execution/session_state.rs#L988-L990

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Closes #11988 .

### CODE REVIEW

File: datafusion/core/src/execution/session_state.rs [hunk 1]
```diff
@@ -1841,4 +1859,68 @@ mod tests {
...
+        let new_state = SessionStateBuilder::new_from_existing(session_state).build();
+        assert!(new_state
+            .catalog_list()
+            .catalog(default_catalog.as_str())
+            .unwrap()
+            .schema(default_schema.as_str())
+            .unwrap()
+            .table_exist("employee"));
+
+        // if `with_create_default_catalog_and_schema` is disabled, the new one shouldn't create default catalog and schema
```
Comment (alamb) [inline datafusion/core/src/execution/session_state.rs:1911]: I ran this test locally without the code changes in this PR and the test fails (as expected) thus verifying test coverage for the new code 👍 

```

assertion failed: new_state.catalog_list().catalog(default_catalog.as_str()).unwrap().schema(default_schema.as_str()).unwrap().table_exist("employee")
thread 'execution::session_state::tests::test_from_existing' panicked at datafusion/core/src/execution/session_state.rs:1887:9:
assertion failed: new_state.catalog_list().catalog(default_catalog.as_str()).unwrap().schema(default_schema.as_str()).unwrap().table_exist("employee")
stack backtrace:
   0: rust_begin_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panicking.rs:146:5
   3: datafusion::execution::session_state::tests::test_from_existing
             at ./src/execution/session_state.rs:1887:9
   4: datafusion::execution::session_state::tests::test_from_existing::{{closure}}
             at ./src/execution/session_state.rs:1848:32
   5: core::ops::function::FnOnce::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:250:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Review by alamb — APPROVED** [review]
Thanks @goldmedal  -- this makes sense to me

**Comment by alamb** [comment]
Thanks again @goldmedal

**Comment by goldmedal** [comment]
Thanks @alamb ❤️

**Pull request merged**